src/dfx/src/commands/deps/pull.rs

```diff
@@ -0,0 +1,340 @@
+        Ok(Some(wasm_hash_raw)) => {
+            let wasm_hash_str = String::from_utf8(wasm_hash_raw)?;
+            trace!(
+                logger,
+                "Canister {canister_id} specified a custom hash: {wasm_hash_str}"
+            );
+            hex::decode(wasm_hash_str)?
+        }
+        Ok(None) => {
+            let canister_status =
+                get_canister_status(env, canister_id, &CallSender::SelectedId).await?;
+            match canister_status.module_hash {
+                Some(hash_on_chain) => hash_on_chain,
+                None => {
+                    bail!("Canister {canister_id} doesn't have module hash. Perhaps it's not installed.");
+                }
+            }
+        }
+        Err(e) => {
+            bail!(e);
+        }
+    };
+
+    pulled_canister.wasm_hash = hex::encode(&hash_on_chain);
+
+    // will save wasm and candid in $(cache_root)/pulled/{canister_id}/
+    let canister_dir = get_cache_root()?
+        .join("pulled")
+        .join(canister_id.to_string());
+    std::fs::create_dir_all(&canister_dir)
+        .with_context(|| format!("Failed to create dir at {:?}", &canister_dir))?;
+
+    let wasm_path = get_pulled_wasm_path(canister_id)?;
+
+    // skip download if cache hit
+    let mut cache_hit = false;
+    let wasm_url_txt_path = get_wasm_url_txt_path(canister_id)?;
+    if wasm_path.exists() {
+        let bytes = std::fs::read(&wasm_path)?;
+        let hash_cache = Sha256::digest(bytes);
+        if hash_cache.as_slice() == hash_on_chain {
+            cache_hit = true;
+            trace!(logger, "The canister wasm was found in the cache.");
+            let wasm_url_str = std::fs::read_to_string(&wasm_url_txt_path)
+                .with_context(|| format!("Failed to read {:?}", &wasm_url_txt_path))?;
+            pulled_canister.wasm_url = Some(wasm_url_str);
+        }
+    }
+    if !cache_hit {
+        // fetch `dfx:wasm_url`
+        let wasm_url_raw = fetch_metatdata(agent, canister_id, DFX_WASM_URL)
+            .await?
+            .ok_or_else(|| {
+                anyhow!("`{DFX_WASM_URL}` metadata not found in canister {canister_id}.")
+            })?;
+        let wasm_url_str = String::from_utf8(wasm_url_raw)?;
+        let wasm_url = reqwest::Url::parse(&wasm_url_str)?;
+        write_to_tempfile_then_rename(wasm_url_str.as_bytes(), &wasm_url_txt_path)?;
+        pulled_canister.wasm_url = Some(wasm_url_str);
+
+        // download
+        let response = reqwest::get(wasm_url.clone()).await?;
+        let status = response.status();
+        if status.is_client_error() || status.is_server_error() {
+            bail!("Failed to download wasm from url: {wasm_url}.\n  StatusCode: {status}");
+        }
+        let content = response.bytes().await?;
+
+        // hash check
+        let mut hasher = Sha256::new();
+        hasher.update(&content);
+        let hash_download = hasher.finalize();
+        if hash_download.as_slice() != hash_on_chain {
```

Comment:
Should this take into account `DFX_WASM_URL` if present?

Reply:
Could you elaborate this a bit?

The code just downloaded from `DFX_WASM_URL` above.

Reply:
https://github.com/dfinity/sdk/blob/6e033c4f17deae2ed659d6d98fedc515eac31d83/src/dfx/src/commands/deps/pull.rs#L153-L171

`hash_on_chain` is from above code which took `DFX_WASM_HASH` into account.

Reply:
Ah, now I see.  I confused by this:
> get the hash of the on chain canister
Suggestion: rename `hash_on_chain` to `expected_hash`.

While technically they are both "on chain" (either the `dfx:wasm_hash` hash from metadata, or the `module_hash`), they are "on chain" in different ways.